sensors/trigger.go

```diff
@@ -30,60 +31,126 @@ import (
 	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
 )
 
-// processTriggers checks if all event dependencies are complete and then starts executing triggers
-func (sec *sensorExecutionCtx) processTriggers() {
-	// to trigger the sensor action/s we need to check if all event dependencies are completed and sensor is active
-	if sec.sensor.AreAllNodesSuccess(v1alpha1.NodeTypeEventDependency) {
-		sec.log.Info().Msg("all event dependencies are marked completed, processing triggers")
+// canProcessTriggers evaluates whether triggers can be processed and executed
+func (sec *sensorExecutionCtx) canProcessTriggers() (bool, error) {
+	if sec.sensor.Spec.DependencyGroups != nil {
+		groups := make(map[string]interface{}, len(sec.sensor.Spec.DependencyGroups))
+	group:
+		for _, group := range sec.sensor.Spec.DependencyGroups {
+			for _, dependency := range group.Dependencies {
+				if nodeStatus := sn.GetNodeByName(sec.sensor, dependency); nodeStatus.Phase != v1alpha1.NodePhaseComplete {
+					groups[group.Name] = false
+					continue group
+				}
+			}
+			sn.MarkNodePhase(sec.sensor, group.Name, v1alpha1.NodeTypeDependencyGroup, v1alpha1.NodePhaseComplete, nil, &sec.log, "dependency group is complete")
+			groups[group.Name] = true
+		}
+
+		expression, err := govaluate.NewEvaluableExpression(sec.sensor.Spec.Circuit)
+		if err != nil {
+			return false, fmt.Errorf("failed to create circuit expression. err: %+v", err)
+		}
 
-		// labels for K8s event
-		labels := map[string]string{
-			common.LabelSensorName: sec.sensor.Name,
-			common.LabelOperation:  "process_triggers",
+		result, err := expression.Evaluate(groups)
+		if err != nil {
+			return false, fmt.Errorf("failed to evaluate circuit expression. err: %+v", err)
 		}
 
-		for _, trigger := range sec.sensor.Spec.Triggers {
-			if err := sec.executeTrigger(trigger); err != nil {
-				sec.log.Error().Str("trigger-name", trigger.Name).Err(err).Msg("trigger failed to execute")
+		return result == true, err
+	}
 
-				sn.MarkNodePhase(sec.sensor, trigger.Name, v1alpha1.NodeTypeTrigger, v1alpha1.NodePhaseError, nil, &sec.log, fmt.Sprintf("failed to execute trigger. err: %+v", err))
+	if sec.sensor.AreAllNodesSuccess(v1alpha1.NodeTypeEventDependency) {
+		return true, nil
+	}
 
-				// escalate using K8s event
-				labels[common.LabelEventType] = string(common.EscalationEventType)
-				if err := common.GenerateK8sEvent(sec.kubeClient, fmt.Sprintf("failed to execute trigger %s", trigger.Name), common.EscalationEventType,
-					"trigger failure", sec.sensor.Name, sec.sensor.Namespace, sec.controllerInstanceID, sensor.Kind, labels); err != nil {
-					sec.log.Error().Err(err).Msg("failed to create K8s event to escalate trigger failure")
-				}
-				continue
+	return false, nil
+}
+
+// canExecuteTrigger determines whether a trigger is executable based on condition set on trigger
+func (sec *sensorExecutionCtx) canExecuteTrigger(trigger v1alpha1.Trigger) bool {
+	if trigger.When == nil {
```

Comment:
so does this make this change non-breaking? for existing sensors defined without groups or `when` this will still allow the trigger to fire?